log4j-api/src/main/java/org/apache/logging/log4j/util/ServiceLoaderUtil.java

```diff
@@ -60,43 +80,48 @@ private ServiceLoaderUtil() {
-            try {
-                @SuppressWarnings("unchecked")
-                final Class<T> serviceClass = (Class<T>) Class.forName(serviceClassName);
-                if (serviceType.isAssignableFrom(serviceClass)) {
-                    return LoaderUtil.newInstanceOf(serviceClass);
-                }
-                getLogger().error("Unable to load service {}: it is not of the required type {}", serviceClass,
-                        serviceType);
-            } catch (ClassNotFoundException | InstantiationException | IllegalAccessException
-                    | InvocationTargetException e) {
-                getLogger().error("Unable to load service class {}", serviceClassName, e);
+        try {
+            final T service = LoaderUtil.newCheckedInstanceOfProperty(serviceType.getName(), serviceType);
+            if (service != null) {
+                return Optional.of(service);
             }
+        } catch (ReflectiveOperationException | ClassCastException e) {
+            logger.error("Unable to load service class {}", serviceType.getName(), e);
         }
         // 2. Use the ServiceLoader:
-        final List<T> services = loadServices(serviceType,
-                t -> getLogger().info("Unable to load service class for service {}", serviceType.getName(), t));
+        final List<T> services = loadServices(serviceType, classLoaders, true).collect(Collectors.toList());
         if (services.size() > 0) {
             if (services.size() > 1) {
-                getLogger().warn("Multiple versions of service {} are available on the classpath: {}", serviceType,
+                logger.warn("Multiple versions of service {} are available on the classpath: {}", serviceType,
                         services);
             }
-            return services.get(0);
+            return Optional.of(services.get(0));
         }
-        return null;
+        return Optional.empty();
     }
 
-    static <T> List<T> loadServices(final Class<T> serviceType, final Consumer<Throwable> logger) {
-        final List<T> services = new ArrayList<T>();
-        for (final ClassLoader loader : LoaderUtil.getClassLoaders()) {
-            addServices(serviceType, loader, logger, services);
-        }
+    static <T> Stream<T> loadServices(final Class<T> serviceType, ClassLoader[] classLoaders, final boolean verbose) {
+        final Stream<T> services = Arrays.stream(classLoaders)
+                .flatMap(classLoader -> loadClassloaderServices(serviceType, classLoader, verbose));
         if (USE_HK2_SERVICE_LOCATOR) {
-            addOsgiServices(serviceType, services);
+            try {
+                return Stream.concat(services, loadOsgiServices(serviceType));
+            } catch (Throwable e) {
+                if (verbose) {
+                    StatusLogger.getLogger().error("Unable to load OSGI services for service {}", serviceType, e);
+                }
+            }
         }
         return services;
```

Comment:
I wonder if we want something like `return services.unique()` here to avoid loading the same instance from multiple related classloaders?

Reply:
`.distinct()` rather (a real stream method!)

Reply:
I'll go for a single instance per class for now. The `Provider` service overrides `equals/hashCode`, but the other service does not.